Fix `bundler/inline` not properly installing gems with extensions when used more than once

tbates-redarc opened the pull request:
## What was the end-user or developer problem that led to this PR?

The following test script fails on Windows (RubyInstaller) with Ruby 3.1 and Ruby 2.6 using Bundler 2.4.5:

```ruby
require 'bundler/inline'
gemfile(true) do
  source 'http://rubygems.org'
  gem 'bundler'
end
gemfile(true) do
  source 'http://rubygems.org'
  gem 'racc'
end
```

<details><summary>The output is as follows:</summary>

```
Resolving dependencies...
Using bundler 2.4.5
Fetching gem metadata from http://rubygems.org/.
Resolving dependencies...
Using bundler 2.4.5
Installing racc 1.6.2 with native extensions
C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer/parallel_installer.rb:189:in `handle_error': Gem::Ext::BuildError: ERROR: Failed to build gem native extension. (Bundler::InstallError)

    current directory: C:/Ruby31-x64/lib/ruby/gems/3.1.0/gems/racc-1.6.2/ext/racc/cparse
C:/Ruby31-x64/bin/ruby.exe -I C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0 extconf.rb
C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/definition.rb:35:in `build': C:/Ruby31-x64/lib/ruby/gems/3.1.0/gems/racc-1.6.2/ext/racc/cparse/Gemfile not found (Bundler::GemfileNotFound)
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler.rb:216:in `definition'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler.rb:164:in `setup'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/setup.rb:20:in `block in <top (required)>'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/ui/shell.rb:159:in `with_level'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/ui/shell.rb:111:in `silence'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/setup.rb:20:in `<top (required)>'
        from <internal:C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from <internal:C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems.rb:1360:in `<top (required)>'
        from <internal:gem_prelude>:2:in `require'
        from <internal:gem_prelude>:2:in `<internal:gem_prelude>'

extconf failed, exit code 1

Gem files will remain installed in C:/Ruby31-x64/lib/ruby/gems/3.1.0/gems/racc-1.6.2 for inspection.
Results logged to C:/Ruby31-x64/lib/ruby/gems/3.1.0/extensions/x64-mingw-ucrt/3.1.0/racc-1.6.2/gem_make.out

  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/ext/builder.rb:102:in `run'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/ext/ext_conf_builder.rb:28:in `build'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/ext/builder.rb:171:in `build_extension'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/ext/builder.rb:205:in `block in build_extensions'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/ext/builder.rb:202:in `each'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/ext/builder.rb:202:in `build_extensions'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/rubygems/installer.rb:843:in `build_extensions'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/rubygems_gem_installer.rb:72:in `build_extensions'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/rubygems_gem_installer.rb:28:in `install'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/source/rubygems.rb:200:in `install'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer/gem_installer.rb:54:in `install'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer/parallel_installer.rb:155:in `do_install'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer/parallel_installer.rb:146:in `block in worker_pool'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/worker.rb:62:in `apply_func'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/worker.rb:57:in `block in process_queue'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/worker.rb:54:in `loop'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/worker.rb:54:in `process_queue'
  C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/worker.rb:90:in `block (2 levels) in create_threads'

An error occurred while installing racc (1.6.2), and Bundler cannot continue.

In Gemfile:
  racc
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer/parallel_installer.rb:96:in `call'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer/parallel_installer.rb:67:in `call'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer.rb:244:in `install_in_parallel'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer.rb:201:in `install'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer.rb:89:in `block in run'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/process_lock.rb:12:in `block in lock'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/process_lock.rb:9:in `open'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/process_lock.rb:9:in `lock'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer.rb:71:in `run'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/installer.rb:23:in `install'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/inline.rb:58:in `block (2 levels) in gemfile'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/settings.rb:131:in `temporary'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/inline.rb:57:in `block in gemfile'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/settings.rb:131:in `temporary'
        from C:/Ruby31-x64/lib/ruby/site_ruby/3.1.0/bundler/inline.rb:51:in `gemfile'
        from test.rb:9:in `<main>'
```

</details>

## What is your fix for the problem, implemented in this PR?

 Wrapping the body of the `gemfile` method in `Bundler.with_unbundled_env` fixed it.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

deivid-rodriguez commented:
I tend to recall setting `BUNDLE_GEMFILE` to the empty string instead of `nil` was an intentional choice to avoid Bundler to default Bundler mechanism to infer the Gemfile automatically when `BUNDLE_GEMFILE` not set. It would be good to double check git/issue history for the whole story.

Also, you may want to add a CI job that runs on Ruby 2.6 on Windows to get this to reproduce. You can look at `.github/workflows/install-rubygems.yml` for inspiration.

tbates-redarc commented:
You're right, it was in rubygems/bundler#7537. Reverted the changes to the BUNDLER_GEMFILE test.

The issue is reproducible on Ruby 3.1 so I have added a test for it but not changed the workflow.

deivid-rodriguez reviewed (approved):
Thanks so much for your improvements to `bundler/inline` 😍

ChrisBr commented:
Note that if you set e.g. auth env variables (e.g. `BUNDLE_MY__PRIVATE__REGISTRY__IO`) these are ignored now. Is this an expected side effect?

deivid-rodriguez commented:
I'd say not expected, does using `Bundler.with_original_env` instead fix the problem?

ChrisBr commented:
This is what we use at the moment

```ruby
require "bundler/inline"

gemfile do
  source "https://rubygems.org"
  gem "foo", source: "https://my.private.registry.io/"
end
```

I haven't tried it but would be surprised if it worked with `Bundler.with_original_env` because the block is wrapped into `Bundler.with_unbundled_env` inside `gemfile`. Or did you mean to replace `Bundler.with_unbundled_env` with `Bundler.with_original_env` inside `gemfile`?

deivid-rodriguez commented:
Yes, I meant replacing it.
